src/main/kotlin/com/vanniktech/maven/publish/MavenPublishConfigurer.kt

```diff
@@ -0,0 +1,116 @@
+import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.publish.maven.plugins.MavenPublishPlugin as GradleMavenPublishPlugin
+import org.gradle.api.tasks.bundling.AbstractArchiveTask
+import org.gradle.plugins.signing.SigningPlugin
+import java.net.URI
+
+internal class MavenPublishConfigurer(private val project: Project) : Configurer {
+
+  private val publication: MavenPublication
+
+  init {
+    project.plugins.apply(GradleMavenPublishPlugin::class.java)
+    project.plugins.apply(SigningPlugin::class.java)
+
+    val publications = project.publishing.publications
+    publication = publications.create("maven", MavenPublication::class.java) { publication ->
+      val publishPom = MavenPublishPom.fromProject(project)
+
+      publication.groupId = publishPom.groupId
+      publication.artifactId = publishPom.artifactId
+      publication.version = publishPom.version
+
+      publication.pom { pom ->
+        pom.name.set(publishPom.name)
+        pom.packaging = publishPom.packaging
+        pom.description.set(publishPom.description)
+        pom.url.set(publishPom.url)
+
+        pom.scm {
+          it.url.set(publishPom.scmUrl)
+          it.connection.set(publishPom.scmConnection)
+          it.developerConnection.set(publishPom.scmDeveloperConnection)
+        }
+
+        pom.licenses { licenses ->
+          licenses.license {
+            it.name.set(publishPom.licenseName)
+            it.url.set(publishPom.licenseUrl)
+            it.distribution.set(publishPom.licenseDistribution)
+          }
+        }
+
+        pom.developers { developers ->
+          developers.developer {
+            it.id.set(publishPom.developerId)
+            it.name.set(publishPom.developerName)
+          }
+        }
+      }
+    }
+
+    project.signing.apply {
+      setRequired(project.isSigningRequired)
+      sign(publication)
+    }
+  }
+
+  override fun configureTarget(target: MavenPublishTarget) {
+    project.publishing.repositories.maven { repo ->
+      repo.name = target.repositoryName
+      repo.url = target.repositoryUrl(project.version.toString())
+      if (target.repositoryUsername != null) {
+        repo.credentials { it ->
+          it.username = target.repositoryUsername
+          it.password = target.repositoryPassword
+        }
+      }
+    }
+
+    // create task that depends on new publishing task for compatibility and easier switching
+    project.tasks.register(target.taskName) {
+      val publicationTaskName = publishTaskName(target.repositoryName)
+      it.dependsOn(project.tasks.named(publicationTaskName))
+    }
+  }
+
+  private val MavenPublishTarget.repositoryName get(): String {
+    return when (name) {
+      DEFAULT_TARGET -> "maven"
+      LOCAL_TARGET -> "local"
+      else -> name
+    }
+  }
+
+  private fun MavenPublishTarget.repositoryUrl(version: String): URI {
+    val url = if (version.endsWith("SNAPSHOT")) {
+      snapshotRepositoryUrl ?: releaseRepositoryUrl
```

Comment:
Why do we default to release url here?

Reply:
To mirror the `uploadArchives` behavior. When no snapshot repository is specified it will always upload to the non snapshot repository. This is also how `installArchives` works which doesn't have a separate snapshot repository.

We could make it more explicit and require a `snapshotRepositoryUrl` for snapshots. For `installArchives` we'd then just set it to the same path.